core/Pos/Binary/Class/Core.hs

```diff
@@ -1,171 +1,589 @@
+                  1 -> do !x <- decode
+                          return (Just x)
+                  _ -> fail $ "decode@Maybe: unknown tag " <> show n
+
+encodeContainerSkel :: (Word -> E.Encoding)
+                    -> (container -> Int)
+                    -> (accumFunc -> E.Encoding -> container -> E.Encoding)
+                    -> accumFunc
+                    -> container
+                    -> E.Encoding
+encodeContainerSkel encodeLen size foldr f  c =
+    encodeLen (fromIntegral (size c)) <> foldr f mempty c
+{-# INLINE encodeContainerSkel #-}
+
+decodeContainerSkelWithReplicate
+  :: Bi a
+  => D.Decoder s Int
+     -- ^ How to get the size of the container
+  -> (Int -> D.Decoder s a -> D.Decoder s container)
+     -- ^ replicateM for the container
+  -> ([container] -> container)
+     -- ^ concat for the container
+  -> D.Decoder s container
+decodeContainerSkelWithReplicate decodeLen replicateFun fromList = do
+    -- Look at how much data we have at the moment and use it as the limit for
+    -- the size of a single call to replicateFun. We don't want to use
+    -- replicateFun directly on the result of decodeLen since this might lead to
+    -- DOS attack (attacker providing a huge value for length). So if it's above
+    -- our limit, we'll do manual chunking and then combine the containers into
+    -- one.
+    size <- decodeLen
+    limit <- D.peekAvailable
+    if size <= limit
+       then replicateFun size decode
+       else do
+           -- Take the max of limit and a fixed chunk size (note: limit can be
+           -- 0). This basically means that the attacker can make us allocate a
+           -- container of size 128 even though there's no actual input.
+           let chunkSize = max limit 128
+               (d, m) = size `divMod` chunkSize
+               buildOne s = replicateFun s decode
+           containers <- sequence $ buildOne m : replicate d (buildOne chunkSize)
+           return $! fromList containers
+{-# INLINE decodeContainerSkelWithReplicate #-}
+
+encodeMapSkel :: (Bi k, Bi v)
+              => (m -> Int)
+              -> ((k -> v -> E.Encoding -> E.Encoding) -> E.Encoding -> m -> E.Encoding)
+              -> m
+              -> E.Encoding
+encodeMapSkel size foldrWithKey =
+  encodeContainerSkel
+    E.encodeMapLen
+    size
+    foldrWithKey
+    (\k v b -> encode k <> encode v <> b)
+{-# INLINE encodeMapSkel #-}
+
+decodeMapSkel :: (Bi k, Bi v) => ([(k,v)] -> m) -> D.Decoder s m
+decodeMapSkel fromList = do
+  n <- D.decodeMapLen
+  let decodeEntry = do
+        !k <- decode
+        !v <- decode
+        return (k, v)
+  fmap fromList (replicateM n decodeEntry)
+{-# INLINE decodeMapSkel #-}
+
+instance (Hashable k, Ord k, Bi k, Bi v) => Bi (HM.HashMap k v) where
+  encode = encodeMapSkel HM.size $ \f acc ->
+      -- We need to encode the list with keys sorted in ascending order as
+      -- that's the only representation we accept during decoding.
```

Comment:
I think the general idea here is that canonicity means essentially we want the decoding process to be injective, so that you cannot possibly get 2 different decoded object starting from the same input. Now, I suspect Andrzej is not enforcing anything here in the decoding side because it's not necessary; sorting everything with `sortWith` during encoding guarantees canonicity by definition. If we check what the official [spec](https://tools.ietf.org/html/rfc7049#section-3.9) has to say, I think it mentions explicitly an `encoder` for the same reasoning:

>  If a protocol considers "canonical" to mean that two encoder
   implementations starting with the same input data will produce the
   same CBOR output, the following four rules would suffice[...]

What do you reckon?